src/alephbet.coffee

```diff
@@ -59,25 +61,30 @@ class AlephBet
       return unless variant
       @storage().set("#{@options.name}:#{goal_name}", true) if props.unique
       utils.log("experiment: #{@options.name} variant:#{variant} goal:#{goal_name} complete")
-      @tracking().goal_complete(@options.name, variant, goal_name)
+      @tracking().goal_complete(this, variant, goal_name, props)
 
     get_stored_variant: ->
       @storage().get("#{@options.name}:variant")
 
     pick_variant: ->
       partitions = 1.0 / @variant_names.length
-      chosen_partition = Math.floor(Math.random() / partitions)
+      chosen_partition = Math.floor(@_random('variant') / partitions)
       variant = @variant_names[chosen_partition]
       utils.log("#{variant} picked")
       variant
 
     in_sample: ->
       active = @storage().get("#{@options.name}:in_sample")
       return active unless typeof active is 'undefined'
-      active = Math.random() <= @options.sample
+      active = @_random('sample') <= @options.sample
       @storage().set("#{@options.name}:in_sample", active)
       active
 
+    _random: (salt) ->
+      return utils.random() unless @user_id
+      seed = "#{@name}.#{salt}.#{@user_id}"
+      utils.random(seed)
```

Comment:
Take a look at the implementation. It's super small and simple. It uses a hash function with the seed and generates a random number based on the seed. If the seed is different for different users, the result hash will produce a different hash, which would get "translated" to a different number between 0 and 1. So for a given experiment, salt and user_id, we'll get a pseudo-random number. If you send the same user_id (or `null`), then you'll always get the same result, which won't be (pseudo)random...